Multistage compatibility for xhatshufflelooper spoke

sylvainczx opened the pull request:
1.  Changing the xhatshufflelooper spoke to make it compatible with multistage problems:
- For multistage problems, the spoke shuffles the scenarios, and then loops over the shuffled list (with an increasing starting point) to fill the scenario tree with feasible policies.
- Two attributes ensures a better estimation: reverse (looping over the reversed shuffled list after, to take into account "hidden" scenarios) and iter_step (increasing the starting point between 2 iterations of more than 1, ensuring a better diversity for the first iterations)
2. Adaptation of baseparsers and vanilla to this.
3. Correcting minor typos

bknueven reviewed (commented):
I left a few minor comments. Overall it seems reasonable.

It would be nice if the `hydro` example was also modified, and perhaps the default settings in baseparsers (e.g., set the default for `with_xhatspecific` to `False`). There's no reason we shouldn't be pointing potential multistage users towards this new capability.

sylvainczx commented on mpisppy/cylinders/xhatshufflelooper_bounder.py:
```diff
@@ -168,37 +179,122 @@ def _vb(msg):
+        #Updating scenarios
+        self._cur_ROOTscen = self._shuffled_snames[self._cycle_idx]
+        self.create_nodescen_dict()
+        #TODO Make a better update system for multistage
+        # if self._multi:
+        #     self.nodescen_dict = {'ROOT': self._cur_ROOTscen}
+        # else:
+        #     self.create_nodescen_dict()
```
I modified ScenarioCycler to change this todo into something done, normally decreasing execution time of an item of the loop.

sylvainczx commented:
I updated, following @bknueven comments. hydro_cylinders is now using xhatshuffle instead of xhatspecific, and with_xhatspecific is set default to False.
Also changed a little bit ScenarioCycler (it was not handling correctly reverse).

bknueven commented on examples/hydro/hydro_cylinders.py:
```diff
@@ -71,23 +71,18 @@ def main():
                                                         all_nodenames,
                                                         BFs,
                                                         scenario_creator_kwargs=scenario_creator_kwargs)
 
```
```suggestion
xhatsuffle_spoke["opt_kwargs"]["all_nodenames"] = all_nodenames

```
I'm a bit surprised this works without this. For a separate PR, we should figure out the `beans` issue with `all_nodenames`

sylvainczx replied:
This is related to the way vanilla treats branching factors and all_nodenames. I made a comment about this in vanilla

sylvainczx commented on mpisppy/utils/vanilla.py:
```diff
@@ -287,6 +289,15 @@ def xhatshuffle_spoke(
+    if _hasit(args,"add_reversed_shuffle"):
+        xhat_options["xhat_looper_options"]["xhatshuffle_iter_step"] = args.xhatshuffle_iter_step
+    
+    if branching_factors:
+        xhat_options["branching_factors"] = branching_factors
+        if all_nodenames is None:
+            all_nodenames = sputils.create_nodenames_from_BFs(branching_factors)
```
@bknueven This is related to your last command. If a branching factor is specified without all_nodenames, the spoke creates the all_nodenames

bknueven replied:
Ah, that makes good sense

bknueven commented on mpisppy/cylinders/xhatshufflelooper_bounder.py:
```diff
@@ -168,37 +179,144 @@ def _vb(msg):
+    
+    def _begin_reverse_epoch(self):
+        self._reversed = True
+        self._shuffled_snames = list(reversed([s[1] for s in self._shuffled_scenarios]))
+        self._original_order = list(reversed([s[0] for s in self._shuffled_scenarios]))
```
```suggestion
        self._shuffled_snames = [s[1] for s in reversed(self._shuffled_scenarios)]
        self._original_order = [s[0] for s in reversed(self._shuffled_scenarios)]
```

bknueven reviewed (approved):
Having tried this out locally with hydro, I think this is an excellent contribution.

I left a final nit in the comments.

@DLWoodruff this does ask if we can re-introduce the "try best first" logic into the shuffle looper. Looking at the logs for the hydro example, the outer scenarios in the tree generally don't do very well. Alternatively, maybe there could be a "companion spoke" to the shuffle which keeps a list of the "best so far" and tries them. For another PR, obviously.
